rpc_util.go

```diff
@@ -307,40 +309,45 @@ func encode(c Codec, msg interface{}, cp Compressor, cbuf *bytes.Buffer, outPayl
 			outPayload.Data = b
 			outPayload.Length = len(b)
 		}
+
 		if cp != nil {
+			// pre-alloc place
+			cbuf.Write(make([]byte, 5))
 			if err := cp.Do(cbuf, b); err != nil {
 				return nil, Errorf(codes.Internal, "grpc: error while compressing: %v", err.Error())
 			}
 			b = cbuf.Bytes()
+			length = uint(len(b) - 5)
+		} else {
+			length = uint(len(b))
 		}
-		length = uint(len(b))
 	}
+
 	if length > math.MaxUint32 {
```

Comment:
If the size can be larger than 4GB, we need to add one more byte (or 2 to 3 bits?) in the header for every message sent to save the length of the message. It's that worth?